Rewrite to use Gradle and Chaquopy

dbnicholson opened the pull request:
It's nice that python-for-android can quickly get you an Android app wrapping your python code, but ultimately it's very limiting since it has to mediate access to the native Android build and runtime APIs. This massive PR adds a native Android build using Gradle with the Python integration coming from Chaquopy. For now this uses an unreleased version of Chaquopy, but that can be dropped when the 15.0.1 release is made.

A major goal here was to move as much of the Android integration out of the Python code as possible. Accessing and manipulating the Android components from Python is slow and fragile as it crosses the JNI and introduces reference cycles between the Java and Python runtimes. I believe all of the existing functionality has been ported to Java directly using the appropriate Android APIs, but there may be regressions.

Fixes: #185

dbnicholson committed:
Initial gradle build infrastructure

This begins the process of building the app natively using Gradle
instead of having python-for-android manage the build. It's setup as a
multi-project with the app as its own project since this is the
customary way Android builds are setup.

The app will build successfully, but it doesn't actually do anything
since there aren't any activities or services defined.

dbnicholson committed:
Add initial chaquopy infrastructure

This adds the chaquopy gradle plugin, includes the necessary python
packages, and uses its PyApplication class for the application. It still
doesn't actually do anything since there isn't any code.

dbnicholson committed:
Generate version numbers in gradle

This is a little more complicated since the versions depend on the
downloaded packages and chaquopy is handling that. A task is inserted
after chaquopy extracts the packages and hooked into AGP's `onVariants`.

The version script is changed to output the JSON file used in
Jenkinsfile-upload. Since Gradle uses input and output files to detect
when a task has changed, this also has the benefit of ensuring the
versions are updated whenever packages are extracted.

dbnicholson committed:
Prune packages downloaded by chaquopy

Like was done in the Makefile, we want to strip unneeded files out of
the downloaded packages to minimize the size of the generated assets. To
simplify the process, the pruning is all done from the script that
previously only pruned unneeded locales.

dbnicholson committed:
Add gradle tasks to download various bundles

Use gradle-download-task to download the various bundles needed. They're
not wired up to be run or used yet.

dbnicholson committed:
gradle: Begin main Kolibri webview activity

Add an Activity with a Webview that will host the Kolibri frontend. At
the moment this just opens the loading screen and does nothing else. The
manifest activity attributes were mostly copied verbatim from the p4a
generated manifest. This requires extracting `loading-screen.zip` and
telling AGP about the generated assets directory, which is a bit tricky.

dbnicholson committed:
Add pre-commit configuration for Java and Kotlin

These will be much more heavily used with chaquopy. I chose the AOSP
Java style instead of the regular Google format since we're writing
Android code. While here:

* Update the existing Java code to match the accepted style.
* Add an EditorConfig file is added to hopefully get your editor to more
  closely match the expected style.
* Add pre-commit to the python requirements and run it during Jenkins
  builds.

The github workflow is dropped since it duplicates the work in Jenkins.
Maybe later we'll move everything to github actions, but for now the
real work is in Jenkins.

dbnicholson committed:
Add chaquopy to pyjnius compatibility module

Unfortunately, even 3rd party packages in the Kolibri stack (zeroconf)
expect to use pyjnius when they think they're on Android. Fortunately,
chaquopy's java.jclass has a nearly identical interface to
jnius.autoclass.

dbnicholson committed:
Extract apps bundle and collections into python source directory

The apps bundle and collections files need to be accessible on disk from
the explore plugin code. The most appropriate way to do that would be to
store the zip files as assets and expand them at runtime. However,
chaquopy already does that for all data files, so just put them in the
python package.

Both to ensure they're extracted when importing `kolibri_android` and
because it's more appropriate, put the data files (including the
automatic provision file) in the package directory rather than the root
of the source set.

dbnicholson committed:
Migrate google services configuration to chaquopy

This just requires enabling the Google Services Gradle plugin and moving
the `google-services.json` to the expected location. All it actually
does is convert the JSON fields into XML resource files that can be
consumed by other Google services like Firebase.

dbnicholson committed:
gradle: Add Firebase Analytics

Currently this is simply enabling or disabling analytics based on
whether it's a release build or not. Later the runtime enablement will
be ported from the python code.

dbnicholson committed:
gradle: Add Firebase Crashlytics

Like analytics, currently this is simply enabling or disabling analytics
based on whether it's a release build or not. Later the runtime
enablement will be ported from the python code. This is also missing the
pieces for the native symbol upload (which we never had before, either).

As a bonus, since chaquopy already converts python exceptions its
`PyException` java class far better than we ever did, we can drop all
of our global exception proxying.

dbnicholson committed:
gradle: Port analytics enabling code to Java

There's no need to do this from Python. The logic should be the same
except that the code also tries to read the value from the manifest.

dbnicholson committed:
Move kolibri_android python code to chaquopy source set

This is not wired up yet and will require several changes, but this is
where it will live to be accessed from java. Note that the script entry
points are not moved over as chaquopy allows calling python attributes
rather than running a single python script until the component ends.

dbnicholson committed:
Add logger class to use consistent EndlessKey tag

Having a separate Log tag for each class makes filtering the logs with
`logcat` annoying. Add a Logger utility class that simply wraps the Log
functions to use a consistent tag.

dbnicholson committed:
gradle: Port webview version code to Java

There's no need to do this from Python. In order to use the `Map.of`
static method, the Java compatibility version needs to be bumped to Java
9.

dbnicholson committed:
Remove dead android code

None of this is used anywhere and just causes a bunch of JNI lookups.

dbnicholson committed:
Remove Android 11 external storage workarounds

We're no longer accessing external storage, so the workaround
monkeypatches are unnecessary.

dbnicholson committed:
Drop remoteshell

To my knowledge, no one at Endless has ever used this. It's neat that
you can spin up an SSH server into the app's storage, but I don't think
that's even that useful for debugging an Android app. Furthermore, it
brings in twisted and all of it's dependencies.

dbnicholson committed:
Drop sys.path mangling

None of these are needed anymore:

* The root of the python source set (`SCRIPT_PATH`) will already be in
  the path.
* Kolibri already takes care of adding its' vendored `dist` directory to
  the path.
* `extra-packages` hasn't existed for a long time.

dbnicholson committed:
gradle: Port Kolibri run mode string code to Java

There's no need to do this from Python and that requires cryptography
and openssl dependencies. Instead, we'll do it in Java and pass the run
mode string during Kolibri initialization.

dbnicholson committed:
python: Use EndlessKey for log tag

There's no need to use different log tags for the Java and Python code.
While here, require the tag argument to the handler so we don't have to
fall back to looking it up through the JNI.

dbnicholson committed:
Refactor Kolibri initialization

With chaquopy we can call multiple python methods with arguments. That
allows initializing Kolibri as a separate step with all the inputs
passed from java rather than gathered back across the JNI from python.

In order to allow multiple concurrent Kolibri services (i.e., a server
and worker service), initialization is synchronized across threads and
processes. Similarly, both the Java and Python side keep track of when
they've already been initialized to allow being called multiple times
from the same process.

While here, kill off `StartupState` since it's only remaining purpose is
to log a message.

dbnicholson committed:
python: Rework context access

Use the current `KolibriActivity` class to access the singleton instance
and change the methods so that `get_context` can be used to get whatever
is the best `Context`. The methods now raise exceptions when an instance
isn't available rather than getting null pointer exceptions.

dbnicholson committed:
Clean up share file interface

The intention from the Kolibri side is that this is for sharing files,
not messages. The point of the message is to accompany provide context
in the OS sharing implementation. The only use for the message in this
case is to use it as a title on an activity chooser. See the
`ACTION_SEND`[1] documentation for details.

As it is, none of this is used in Kolibri right now, so this is just
academic.

1. https://developer.android.com/reference/android/content/Intent#ACTION_SEND

dbnicholson committed:
Add file provider class

The previous implementation purely in python would not have worked
because the generic `FileProvider` class does not serve the
`org.endlessos.Key.fileprovider` authority.

Instead, a subclass is created, a `provider` is added to the manifest,
and a convenience static method is added to construct the appropriate
intent from python.

dbnicholson committed:
python: Convert most pyjnius usage to chaquopy

Use chaquopy to access Java classes instead of `jnius.autoclass` and
`jnius.cast`. Chaquopy's import hook[1] is used to import the Java
classes directly. The only remain piece is the `runnable` module, which
will go away later.

1. https://chaquo.com/chaquopy/doc/current/python.html#import-hook

dbnicholson committed:
Add Kolibri service

Run Kolibri as a service which is bound from the main webview activity.
The service is bound and unbound in the activity's start and stop
states. Since the service runs in a separate process, this gives Kolibri
a chance to stop gracefully when the activity is killed. In order for
the activity to find the current URL and app key, a small `Message`
based IPC is used.

Chaquopy expects to return from Python calls, so rather than blocking on
the server, the process bus is started and stopped as the service is
created and destroyed. This requires a bit of refactoring in the python
code.

Similarly, since we have full access to the activity and service
lifecycle hooks, a lot of the python code can go away as it simply needs
to start the Kolibri process bus.

dbnicholson committed:
Support release upload signing

If the upload signing key settings are available in the
`upload.properties` file, use them to sign the release build.

dbnicholson committed:
Simplify containers

With Gradle and Chaquopy we don't need to do any native building. All
that's needed is Java, Python and a few Python packages for our own
scripts.

dbnicholson committed:
Simplify SDK installation

With Chaquopy we don't need the NDK since we use its pre-compiled native
libraries. Furthermore, this drops the p4a specific (and slightly wrong)
`ANDROIDSDK` and `ANDROIDNDK` variables. Instead, just the
`ANDROID_HOME` variable pointing to the SDK directory[1] is needed.

1. https://developer.android.com/tools/variables#envar

dbnicholson committed:
Rework build to use Gradle

Update the Makefile and Jenkinsfile to run Gradle tasks. The Makefile
isn't really needed anymore (besides for installing the SDK), but is
kept for convenience.

dbnicholson committed:
Nuke p4a from orbit

The build now uses Gradle and Chaquopy natively. So long, p4a.

dbnicholson committed:
Drop rundocker script

Although the image is used, we haven't used the script for configuring
and running the build in the container for a long time. While here, also
remove the buildkite configuration, which even LE isn't using anymore.

dbnicholson committed:
Update documentation for use with Gradle

dbnicholson committed:
Stop disabling Gradle daemon in container

We now run Gradle multiple times from the container, so letting it start
a daemon does improve the overall speed. Furthermore, even if you say
not to use a daemon, Gradle 8 starts a single use daemon since it's part
of the implementation. Like the comment previously said, any daemon will
be killed as soon as the container is stopped.

dylanmccall commented:
This is spectacular :star_struck: 

I skimmed through this and it looks good. It's going to take me a while to review properly, but I just wanted to pipe in to say that I looked at it and I am super excited about it :)

I have been using https://github.com/dbnicholson/kolibri-chaquopy for a while for my own little task and it has been working really well (especially after I gave up on running Android Studio on my laptop). Am I correct in assuming this is mostly cloning stuff from there, or are there some other significant changes from two weeks ago?

dylanmccall commented on .pre-commit-config.yaml:
```diff
@@ -28,3 +28,11 @@ repos:
         hooks:
             -   id: reorder-python-imports
+
+    - repo: https://github.com/macisamuele/language-formatters-pre-commit-hooks
+      rev: v2.10.0
+      hooks:
+        - id: pretty-format-java
+          args: [--autofix, --aosp]
+        - id: pretty-format-kotlin
+          args: [--autofix]
```
Ah, I was going to ask about autoformatting Java now all that dirty Python is gone, and there it is! Do you think it could be worth adding something for XML, as well?

dbnicholson replied:
That would be nice. I was looking at https://pre-commit.com/hooks.html and the only thing that really popped out was https://github.com/comkieffer/xml-linter-hook. However, `xmllint --format` isn't that great and the hook actually doesn't support it. I'm all ears if you have any ideas!

dbnicholson committed:
Downgrade to build-tools 33.0.1

As it turns out, the Android Gradle Plugin 8.1 series specifies that it
uses 33.0.1 and tries to install it if needed. That fails in the
containers. We could specify `buildToolsVersion`, but it seems likely to
be forgotten about and the only change in 33.0.2 doesn't seem to affect
us.

dbnicholson commented:
> I have been using https://github.com/dbnicholson/kolibri-chaquopy for a while for my own little task and it has been working really well (especially after I gave up on running Android Studio on my laptop). Am I correct in assuming this is mostly cloning stuff from there, or are there some other significant changes from two weeks ago?

It's almost entirely the same as kolibri-chaquopy with a couple exceptions:

1. In kolibri-chaquopy there are 2 kolibri services - one for the server and one for the worker. I decided to leave that as a single service since that's how it was already, but we can certainly split that up. I also don't know great that is if the worker service is running all the time in a separate process. `WorkManager` would obviously change that.
2. There were several more features here that required additional Java porting.

The 2 things I know that are currently missing are:

1. Animated view restoring from https://phabricator.endlessm.com/T34576. At first I didn't really think it was necessary, and then I wasn't sure I was following the semantics correctly.
2. History clearing from https://phabricator.endlessm.com/T34551. Again, I wasn't totally sure of the semantics.

dbnicholson committed:
Install google_atd image with SDK

The Google automated test device (ATD) images are smaller and designed
just for testing. This is what's configured for use with the
`managedDevices` testing. Without pre-installing it, those tests can't
be run in CI since the job can't install to the SDK.

dbnicholson committed:
Update to kolibri-explore-plugin 7.8.0

dbnicholson commented:
FYI, I added a commit to bump the explore plugin version to 7.8.0. Otherwise, it's the same.

dylanmccall commented on app/src/main/java/org/endlessos/key/KolibriActivity.java:
```diff
@@ -0,0 +1,348 @@
+        Logger.i("Set last URL path to " + lastUrlPath);
+    }
+
+    private void loadLastUrl(Uri serverUrl) {
```
Nit: the name of this method threw me off. ("Why is it taking a URL if it's loading the last URL?"). Maybe `changeBaseUrl()` would describe what's happening here?

dbnicholson replied:
I was thinking about it from the perspective of handling the last URL and I thought I'd changed the name to `loadLastUrlPath` at some point to match the field name. Regardless, I think what you've described makes a little more sense. What the path is and how it's handled are an implementation detail that doesn't need to be reflected in the method name. I'll update.

dbnicholson replied:
I decided to call it `updateServerUrl`

dylanmccall reviewed (approved):
So, yes, I'm happy with this and I think we can go ahead and merge it. It's all so nice and neat and tidy! :)

Let's create a follow-up task where we can evaluate whether we still want to keep (and possibly re-implement) that separate loading screen webview.

I think the stuff about history clearing we can drop altogether (if it was still there?): the problem that addressed is the same as the separate loading webview; just that the loading screen uses the same webview as the main application, and if we change it to be a separate webview, the history stack only has things that a user would expect.

dbnicholson commented:
> Let's create a follow-up task where we can evaluate whether we still want to keep (and possibly re-implement) that separate loading screen webview.
> 
> I think the stuff about history clearing we can drop altogether (if it was still there?): the problem that addressed is the same as the separate loading webview; just that the loading screen uses the same webview as the main application, and if we change it to be a separate webview, the history stack only has things that a user would expect.

Yeah, the history clearing does pretty much go away with a separate webview. The view switching I quickly implemented didn't make things better in terms of UI, but yes that's more for a followup.

One thought I had about the history the would be more involved is that if the server is stopped and started, you'll have a different base URL. Then if you actually want back in the history I presume it would try to use a URL where there wasn't a server anymore. I think you'd have to rewrite the history to use the new server URL in all the old URLs.

dbnicholson commented:
Got an ack from @manuq on Slack. I'm going to squash the last fixup and merge.
